Java sprint6

itsmekatrine opened the pull request:
Доработана функциональность по техническому заданию спринта

Commit:
Сделать историю посещений неограниченной по размеру
Добавить метод void remove(int id) для удаления задачи из просмотра
Разработка алгоритма со связным списком и HashMap

Commit:
Добавить unit-тесты для:
- встроенного связного списка версий, а также операций добавления и удаления
- удаляемых подзадач, они не должны хранить внутри себя старые id
- эпиков, внутри них не должно оставаться неактуальных id подзадач
- неизменяемости полей в задаче

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

itsmekatrine commented:
Не проходит чек по Caused by: java.lang.IllegalStateException: 11:13: mismatched input '(' expecting ';'
в коде всё корректно написано и всё компилируется

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Yang-Pi commented on src/com/yandex/app/service/InMemoryHistoryManager.java:
```diff
@@ -4,41 +4,88 @@
 public class InMemoryHistoryManager implements HistoryManager {
-    private ArrayList<Task> history;
-    private static final int MAX_HISTORY_SIZE = 10;
+    private Map<Integer, Node> map;
```
думаю более подходящим названием поля будет history)

itsmekatrine replied:
Исправлено в коммите 4c4523811a515b9beac298053945ccfea021d9eb

Yang-Pi commented on src/com/yandex/app/service/InMemoryHistoryManager.java:
```diff
@@ -4,41 +4,88 @@
+        int id = task.getId();
+        Node oldNode = map.get(id);
+        if (oldNode != null) {
+            oldNode.task = task;
```
получается если задача уже существовала в мапе, то в конец списка (истории) будет добавлена дублирующая нода. Одна - oldNode, вторая - новая нода в конце списка. Нужно удалить старую из списка

itsmekatrine replied:
Исправлено в комете 0c4075424825ba7827e74a39a30e81fd63a436b2

Yang-Pi commented on src/com/yandex/app/service/InMemoryHistoryManager.java:
```diff
@@ -4,41 +4,88 @@
+            oldNode.task = task;
+        }
+        linkLast(task);
+        map.put(id, tail);
```
лучше складывать в мапу текущую добавляемую задачу, а не хвост списка - вдруг реализация изменится и хвост будет указывать на самом деле на пустое место. Для этого стоит создавать объекты ноды в этом методе, а в метод linkLast передавать уже готовый объект ноды

itsmekatrine replied:
Исправлено в коммите 0c4075424825ba7827e74a39a30e81fd63a436b2

Yang-Pi commented on src/com/yandex/app/service/InMemoryHistoryManager.java:
```diff
@@ -4,41 +4,88 @@
+    public void linkLast(Task task) {
+        Node newNode = new Node(task);
+        if (head == null) {
+            head = tail = newNode;
```
лучше не пользоваться двойным присваиванием, стиль кода должен быть единым, поэтому лучше разместить в 2 строчки

itsmekatrine replied:
Исправлено в коммите 8310eadf003d5a88842e87a82ee4087cba392920

Yang-Pi commented on src/com/yandex/app/service/InMemoryTaskManager.java:
```diff
@@ -161,6 +171,7 @@ public boolean deleteSubtaskById(int id) {
     }
 
+    @Override
     public void deleteSubtaskFromEpic(int id) {
```
при удалении любых задач не забывай удалять задачи и из истории)

itsmekatrine replied:
Исправлено в коммите f419ef6dc48e097b2eb34389861651d5cf175365

Yang-Pi commented on src/com/yandex/app/test/TaskTest.java:
```diff
@@ -2,15 +2,21 @@
-    TaskManager manager = new InMemoryTaskManager();
-    HistoryManager historyManager = new InMemoryHistoryManager();
+    private TaskManager manager;
+    private HistoryManager historyManager = new InMemoryHistoryManager();
```
для инициализации лучше использовать класс-фабрику Managers - он предоставляет методы, которые возвращают реализации нужных интерфейсов

itsmekatrine replied:
Исправлено в коммите 0a905b3285513e376acf402bdcbd4ec753cebab5

Yang-Pi commented on src/com/yandex/app/test/InMemoryTaskManagerTest.java:
```diff
@@ -12,11 +13,11 @@
 import static org.junit.jupiter.api.Assertions.*;
 
 class InMemoryTaskManagerTest {
```
тестов много, молодец, однако еще нужно покрыть тестами реализацию списка (истории), вот какие тестовые кейсы можно написать:

- создайте две задачи, эпик с тремя подзадачами и эпик без подзадач;

- запросите созданные задачи несколько раз в разном порядке;

- после каждого запроса выведите историю и убедитесь, что в ней нет повторов;

- удалите задачу, которая есть в истории, и проверьте, что при печати она не будет выводиться;

- удалите эпик с тремя подзадачами и убедитесь, что из истории удалился как сам эпик, так и все его подзадачи.

itsmekatrine replied:
Добавлено в коммите 41a1894fd6e482aa0464b6b53cf027b0d84449b6

Yang-Pi commented on src/com/yandex/app/test/EpicTest.java:
```diff
@@ -3,22 +3,31 @@
 class EpicTest {
+    TaskManager manager;
+    private HistoryManager historyManager = new InMemoryHistoryManager();
+    Epic epic;
```
а почему остальные поля не private ?

itsmekatrine replied:
Исправлено в коммите a1a633c9553da76df11d2d2c9e8c6d2fc24247f1

Yang-Pi commented on src/com/yandex/app/service/InMemoryHistoryManager.java:
```diff
@@ -4,41 +4,88 @@
+
+    // метод для удаления узла
+    public void removeNode(Node node) {
+        if (node.prev != null) {
```
на будущее: в реализации связанных списков удобно пользоваться техникой "пустая голова и пустой хвост" - это когда head и tail ссылаются на объекты Node, имеющие task = null, то есть не содержащие полезной нагрузки. Поэтому и название "пустая". Это очень удобно, когда head и tail никогда не null. Добавление и удаление новых нод сводится к простому присваиванию ссылок, без необходимости проверять на null, пустоту списка. Обход списка будет начинаться с head.next

itsmekatrine replied:
Исправлено в коммите 480552ab453945283f17f07f96dfd1272089ee94

Yang-Pi reviewed (commented):
хорошая работа! оставил несколько замечаний и рекомендаций. Еще обрати пожалуйста внимание, что автотесты на гитхабе все должны проходить)

Commit:
Сделать поля private

Commit:
В TaskTest для инициализации использовать класс-фабрику Managers

Commit:
Исправить двойное присваивание

Commit:
- Cкладывать в мапу текущую добавляемую задачу, а не хвост списка
- В метод linkLast передавать уже готовый объект ноды
- Удалить старую ноду из списка

Commit:
Поменять название map на history

Commit:
При удалении задач добавить удаление их из истории

Commit:
Добавлена техника "пустая голова и пустой хвост"

Commit:
Скорректировать юнит-тесты
Добавить новые юнит-тесты

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Commit:
Исправить CheckStyle для PR

Yang-Pi commented on src/com/yandex/app/service/InMemoryHistoryManager.java:
```diff
@@ -8,31 +8,40 @@
-        if (head == null) {
-            head = tail = newNode;
+    public void linkLast(Node newNode) {
+        if (head.next == tail) {
```
так как ты добавляешь в конец и у тебя есть пустая голова и хвост, проверки на пустоту не нужны, достаточно просто добавить элемент в конец, то есть перепривязять prev и next у хвоста (tail) и назначить tail на новую последнюю ноду

itsmekatrine replied:
Исправлено в коммите 07668e0d4fc0a2a535bdfbac697e56914e478948

Yang-Pi commented on src/com/yandex/app/service/InMemoryHistoryManager.java:
```diff
@@ -41,6 +50,9 @@ public List<Task> getTasks() {
 
     // метод для удаления узла
     public void removeNode(Node node) {
+        if (node == head || node == tail) {
```
в методе removeNode тоже не нужны дополнительные проверки) достаточно написать примерно:

node.getPrev().setNext(node.getNext());
        node.getNext().setPrev(node.getPrev());

itsmekatrine replied:
Исправлено в коммите 0157c1ee955121c4d89b7d5f2134dab495054e0b

Yang-Pi reviewed (approved):
работа принята, при желании можешь поправить замечания, раз взялась за реализацию пустой головы)

Commit:
Убрать проверки на пустоту в методе linkLast()

Commit:
Убрать проверки на пустоту в методе removeNode()
